src/sfall_arrays.cc

```diff
@@ -0,0 +1,697 @@
+    }
+
+    void SetArray(const ProgramValue& key, const ProgramValue& val, bool allowUnset, Program* program)
+    {
+        auto keyEl = ArrayElement { key, program };
+        auto it = std::find_if(pairs.begin(), pairs.end(), [&keyEl](const KeyValuePair& pair) -> bool {
+            return pair.key == keyEl;
+        });
+
+        if (it != pairs.end() && isReadOnly()) {
+            // don't update value of key
+            return;
+        }
+
+        if (allowUnset && !isReadOnly() && val.isInt() && val.asInt() == 0) {
+            // after assigning zero to a key, no need to store it, because "get_array" returns 0 for non-existent keys: try unset
+            if (it != pairs.end()) {
+                pairs.erase(it);
+            }
+        } else {
+            if (it == pairs.end()) {
+                // size check
+                if (size() >= ARRAY_MAX_SIZE) {
+                    return;
+                }
+
+                pairs.push_back(KeyValuePair { std::move(keyEl), ArrayElement { val, program } });
+            } else {
+                auto index = it - pairs.begin();
+                pairs[index].value = ArrayElement { val, program };
+            }
+        }
+    }
+
+    void SetArray(const ProgramValue& key, ArrayElement&& val, bool allowUnset)
+    {
+        assert(false && "This method is not used for associative arrays thus it is not implemented");
+    }
+
+    void ResizeArray(int newLen)
+    {
+        if (newLen == -1 || size() == newLen) {
+            return;
+        }
+
+        // only allow to reduce number of elements (adding range of elements is meaningless for maps)
+        if (newLen >= 0 && newLen < size()) {
+            pairs.resize(newLen);
+        } else if (newLen < 0) {
+            if (newLen < (ARRAY_ACTION_SHUFFLE - 2)) return;
+            MapSort(newLen);
+        }
+    }
+
+    ProgramValue ScanArray(const ProgramValue& value, Program* program)
+    {
+        auto valueEl = ArrayElement { value, program };
+        auto it = std::find_if(pairs.begin(), pairs.end(), [&valueEl](const KeyValuePair& pair) {
+            return pair.value == valueEl;
+        });
+
+        if (it == pairs.end()) {
+            return ProgramValue(-1);
+        }
+
+        return it->key.toValue(program);
+    }
+
+private:
+    struct KeyValuePair {
+        ArrayElement key;
+        ArrayElement value;
+    };
+
+    void MapSort(int type)
+    {
+        bool sortByValue = false;
+        if (type < ARRAY_ACTION_SHUFFLE) {
+            type += 4;
+            sortByValue = true;
+        }
+
+        if (sortByValue) {
+            ListSort(pairs, type, [](const KeyValuePair& a, const KeyValuePair& b) -> bool {
+                return a.value < b.value;
+            });
+        } else {
+            ListSort(pairs, type, [](const KeyValuePair& a, const KeyValuePair& b) -> bool {
+                return a.key < b.key;
+            });
+        }
+    }
+
+    std::vector<KeyValuePair> pairs;
```

Comment:
LOL no `map` for assoc arrays. I think it is fine if we know that arrays are usually small (for example less than 100 or 1000 items). Anyway I do not see any problem here